hazelcast/src/main/java/com/hazelcast/query/impl/AndResultSet.java

```diff
@@ -161,6 +150,28 @@ public void remove() {
 
     @Override
     public int size() {
-        return setSmallest.size();
+        if (size == -1) {
+            int calculatedSize = 0;
+            for (Iterator<QueryableEntry> it = iterator(); it.hasNext(); it.next()) {
+                calculatedSize++;
+            }
+            size = calculatedSize;
+        }
+        return size;
+    }
+
+    /**
+     * @return returns estimated size without calculating the full result set in full-result scan.
+     */
+    public int estimatedSize() {
```

Comment:
Yes, it unfortunately does, but the users get the result as the top-level Set interface